### ISSUE TITLE
Question: reproducing ~9s/iter DDP speed with Qwen3-VL 8-frame multi-image config

### ISSUE THREAD

**Issue by DanielLSM** [robometer/robometer#32]
Hi! First, thanks for making this training path available. I’m trying to reproduce the `~9s/iter` DDP speed noted in commit `c43178d`, and I may be missing an effective config detail.

From the current public script/defaults, it looks like the Qwen3-VL RBM run uses an 8-frame **multi-image** representation with per-frame progress tokens and no resizing. In my repro, that representation is much slower and seems dominated by recomputing frozen Qwen3-VL visual embeddings. I wanted to check whether the fast run used the exact public representation, or whether one of the representation/preprocessing knobs differed.

## What I checked upstream

Checked against `upstream/main` at `a669dffc241d7d76bec12f36efd4084d914d017c`.

Relevant defaults in `robometer/configs/config.yaml` appear to be:

```yaml
model:
  use_per_frame_progress_token: true
  use_multi_image: ${data.use_multi_image}

data:
  max_frames: 16
  resized_height: null
  resized_width: null
  use_multi_image: true

training:
  per_device_train_batch_size: 16
  gradient_checkpointing: true
```

Relevant overrides in `run_scripts/run_robometer.sbatch`:

```bash
export BS=16

uv run accelerate launch --config_file robometer/configs/distributed/ddp.yaml \
  --num_processes=4 \
  train.py \
  data.train_datasets=[rbm-1m-id] \
  data.eval_datasets=[rbm-1m-ood] \
  data.max_frames=8 \
  model.train_progress_head=true \
  model.train_preference_head=true \
  training.per_device_train_batch_size=${BS} \
  training.per_device_eval_batch_size=${BS} \
  training.max_steps=15000
```

I don’t see overrides for:

- `data.use_multi_image`
- `model.use_per_frame_progress_token`
- `data.resized_height`
- `data.resized_width`

So I believe the public recipe’s effective representation is:

```text
data.use_multi_image=true
model.use_per_frame_progress_token=true
data.max_frames=8
data.resized_height=null
data.resized_width=null
training.per_device_train_batch_size=16
```

The collator seems consistent with this too: in multi-image mode each frame is appended as an individual `{type: "image"}` item, `content_extras = {}` is used in that path, and the processor is called with `do_resize=False`.

## What I measured

For the 8-frame multi-image representation, each preference sample contains two trajectories, so a per-rank batch of 16 sends:

```text
16 samples × 2 trajectories × 8 frames = 256 images/rank/step
```

Measured collation shape for that geometry:

```text
input_ids:      [16, 1296]
pixel_values:   [65536, 1536]
image_grid_thw: [256, 3]       # each frame has grid [1, 16, 16]
```

For a smaller single-GPU probe with `B=4`, same representation:

```text
input_ids:      [4, 1296]
pixel_values:   [16384, 1536]
image_grid_thw: [64, 3]
```

Single-GPU forward breakdown with `B=4`, `data.max_frames=8`, multi-image/default representation:

```text
wall_seconds: 64.667
time/rbm_forward: 64.649
submodule_times:
  visual:         62.243
  language_model:  2.306
```

Parameter check from the same probe:

```text
visual_total:         415,347,712
visual_trainable:               0
language_total:     4,021,797,376
language_trainable: 4,021,797,376
```

So the visual tower is frozen, but its embeddings are still recomputed every step and dominate runtime.

Frame-count scaling with `B=4` also points to visual-token volume as the bottleneck:

```text
max_frames=1: repeat forward ~7.45s,  pixel_values=[2048,1536]
max_frames=2: repeat forward ~14.51s, pixel_values=[4096,1536]
max_frames=8: repeat forward ~57.44s, pixel_values=[16384,1536]
```

A 4-GPU DDP + gradient-checkpointing probe using the public 8-frame multi-image semantics reached training without OOM and measured:

```text
time/training_step ~= 260.64s
time/rbm_forward   ~= 258.92s
```

## A lower-token probe gets close to the reported speed class

If I switch to video mode with resized inputs, the tensor geometry is much smaller and the forward time is close to the `~9s/iter` regime.

Overrides:

```text
data.use_multi_image=false
model.use_per_frame_progress_token=false
data.resized_height=128
data.resized_width=128
data.max_frames=8
```

Measured shape/timing with `B=4`:

```text
input_ids:           [4, 337]
pixel_values_videos: [2048, 1536]
video_grid_thw:      [8, 3]       # each trajectory has grid [4, 8, 8]

wall_seconds: 11.306
time/rbm_forward: 11.287
submodule_times:
  visual:          8.365
  language_model:  2.788
```

That seems much more consistent with the `~9s/iter` note, but it changes the representation and disables the per-frame progress token path.

## Questions

Could you help confirm the intended fast recipe?

1. Was the `~9s/iter` number in `c43178d` measured with the current public `run_scripts/run_robometer.sbatch` defaults exactly?
2. If yes, could you share the resolved config and environment/library versions from that run?
3. If not, which knobs differed from the public script/defaults?
   - `data.use_multi_image`
   - `model.use_per_frame_progress_token`
   - `data.resized_height` / `data.resized_width`
   - `data.max_frames`
   - cached/precomputed visual features
   - dataset/image preprocessing resolution
   - Qwen / Transformers / qwen-vl-utils / Unsloth versions
4. Is the comment in `run_robometer.sbatch` saying “2 frames” stale? The script currently sets `data.max_frames=8`.
5. Since `visual_trainable=0`, is the intended full-speed path to cache frozen Qwen3-VL visual embeddings, or to use the lower-token video/resized representation?

## Expected / desired outcome

It would be helpful if one of these were made explicit:

- the public DDP recipe reproduces the `~9s/iter` speed class as-is;
- or the repo documents the exact representation/config used for that speed;
- or the default 8-frame multi-image recipe avoids recomputing frozen visual embeddings every step.

## Environment I used for reproduction

```text
Model: Qwen/Qwen3-VL-4B-Instruct
Distributed mode: DDP, 4 processes, bf16
Gradient checkpointing: true
Hardware: NVIDIA GH200 120GB
PyTorch: 2.9.1+cu128
Transformers: 4.57.6
Unsloth: 2026.4.1
FlashAttention available: true
```

Thanks again — mostly trying to understand whether I’m reproducing the intended representation, or whether the public run script is missing one of the speed-critical overrides.

**Comment by carl-qi** [comment]
I also have this issse. Some benchmarking suggests that qwen_vl_util's vision processor is the bottleneck. I'm getting close to 60s/iter with the current code. (per device batchsize=16, max image =8), which is super slow.

**Comment by jesbu1** [comment]
This is strange, can you try the `fsdp` config? 

Did not experience this on our H200 cluster, in fact DDP was much faster with some recent package updates.

`uv run accelerate launch --config_file robometer/configs/distributed/fsdp.yaml --num_processes=N_GPUS_YOU_HAVE train.py ... `

**Comment by carl-qi** [comment]
I tried with ddp and fsdp, both giving similar speed. 

Once i switched to setting use_multi_image=false + use_per_frame_progress_token=false (all frames packed as a signle video input), it sped up training significantly. i think i'm getting close to the reported speed. However, this results in the video processor potentially downsamples patches.

for context, I'm also using a H200 node with 8 GPUs, + 192v CPUs per node.

**Comment by jesbu1** [comment]
Is `unsloth` working for you? It could be an unsloth issue -- the settings you chose are not optimal for reward prediction unfortunately.

I believe it makes the model only have 1 summary frame as input per video (or a strided set of few tokens) that it uses for reward prediction rather than being able to attend to all tokenized frame patches.

**Comment by carl-qi** [comment]
For me, switching back to use_multi_image=True and use_per_frame_progress_token=True seemed to work too. not sure which change that made it faster. I will update if i find anything.
